In [33]:
import pandas as pd

# Specify your GitHub path
path = "https://raw.githubusercontent.com/alimanager/smartBuilding/master/"

# Get a list of all .csv files in the GitHub directory
# Please note that GitHub API doesn't allow listing directory contents, 
# so you have to provide the names of your csv files manually
csv_files = ['01_occ.csv', '02_win.csv', '03_light.csv','04_plug.csv','05_temp_in.csv','06_rhu_in.csv','07_rad_global.csv','08_temp_out.csv','09_rhu_out.csv','10_wsp.csv','11_wdi.csv']  # Replace with your actual file names

# Create a dictionary to store DataFrames with corresponding names for each .csv file
dfs = {}

# Read each .csv file, rename the DataFrame, and store it in the dictionary
for file_name in csv_files:
    df_name = file_name.replace('.csv', '')  # Extract DataFrame name from the file name
    url = path + file_name
    dfs[df_name] = pd.read_csv(url)  # Create DataFrame with the extracted name

    # Perform the merge based on a common key (e.g., 'common_column')
    # Replace 'common_column' with the actual column name that is common across all DataFrames
    merged_df = dfs['01_occ']  # Initialize merged_df with one of the DataFrames
    for df_name, df in dfs.items():
        if df_name != '01_occ':  # Skip the first DataFrame since it's already stored in merged_df
            merged_df = pd.merge(merged_df, df, on='timestamp [dd/mm/yyyy HH:MM]',how='outer' )
            # merged_df.fillna(method='ffill', inplace=True)
            # Forward-fill missing temperature values
            merged_df.fillna(method='ffill', inplace=True)


    # Now you have a merged DataFrame named 'merged_df' containing data from all .csv files
    nan_df = merged_df.isna()

    # If there are any NaN values, the nan_df DataFrame will contain True in those positions.
    # You can check if there are any NaN values in the entire DataFrame by using the any() method.
    if nan_df.any().any():
        print("The DataFrame contains NaN values.")
   
    merged_df

|      | Le premier événement   | Le dernier événement    |
|:-----|:--------|:--------|
| mean | 7.35 H  | 15.76 H |
| std  | 3.02 H  | 2.48 H  |
| min  | 0.00 H  | 3.00 H  |
| 25%  | 6.00 H  | 15.00 H |
| 50%  | 6.00 H  | 16.00 H |
| 75%  | 7.00 H  | 17.00 H |
| max  | 21.00 H | 23.00 H |

In [26]:
sample_df = merged_df.filter(like=" [0:vacant 1:occupied]")
for i in range(10, 25):
    random_column = np.random.choice(sample_df.columns)
    mean = np.mean(sample_df[random_column])
    std = np.std(sample_df[random_column])
    # sample_df[f'o{i} [0:vacant 1:occupied]'] = sample_df[random_column].sample(len(sample_df), replace=True).values
    # sample_df[f'o{i} [0:vacant 1:occupied]'] = np.random.normal(mean, std, len(sample_df))
    sample_df[f'o{i} [0:vacant 1:occupied]'] = np.random.beta(2, 2, len(sample_df))

sample_df

C:\Users\Amghar\AppData\Local\Temp\ipykernel_56204\3952470574.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sample_df[f'o{i} [0:vacant 1:occupied]'] = np.random.beta(2, 2, len(sample_df))
C:\Users\Amghar\AppData\Local\Temp\ipykernel_56204\3952470574.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sample_df[f'o{i} [0:vacant 1:occupied]'] = np.random.beta(2, 2, len(sample_df))
C:\Users\Amghar\AppData\Local\Temp\ipykernel_56204\3952470574.py:8: SettingWithCopyWarning: 
A value is trying to be set on 

ki [0:vacant 1:occupied]  o1_1 [0:vacant 1:occupied]  \
0                           0.0                         0.0   
1                           0.0                         0.0   
2                           0.0                         0.0   
3                           0.0                         0.0   
4                           0.0                         0.0   
...                         ...                         ...   
35035                       0.0                         0.0   
35036                       0.0                         0.0   
35037                       0.0                         0.0   
35038                       0.0                         0.0   
35039                       0.0                         0.0   

       o1_2 [0:vacant 1:occupied]  o1_3 [0:vacant 1:occupied]  \
0                             0.0                         0.0   
1                             0.0                         0.0   
2                             0.0                         0.0   
3                             0.0                         0.0   
4                             0.0                         0.0   
...                           ...                         ...   
35035                         0.0                         0.0   
35036                         0.0                         0.0   
35037                         0.0                         0.0   
35038                         0.0                         0.0   
35039                         0.0                         0.0   

       o1_4 [0:vacant 1:occupied]  o1_5 [0:vacant 1:occupied]  \
0                             0.0                         0.0   
1                             0.0                         0.0   
2                             0.0                         0.0   
3                             0.0                         0.0   
4                             0.0                         0.0   
...                           ...                         ...   
35035                         0.0                         0.0   
35036                         0.0                         0.0   
35037                         0.0                         0.0   
35038                         0.0                         0.0   
35039                         0.0                         0.0   

       o2 [0:vacant 1:occupied]  o3 [0:vacant 1:occupied]  \
0                           0.0                       0.0   
1                           0.0                       0.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
35035                       0.0                       0.0   
35036                       0.0                       0.0   
35037                       0.0                       0.0   
35038                       0.0                       0.0   
35039                       0.0                       0.0   

       o4 [0:vacant 1:occupied]  o10 [0:vacant 1:occupied]  ...  \
0                           0.0                   0.460766  ...   
1                           0.0                   0.481943  ...   
2                           0.0                   0.325504  ...   
3                           0.0                   0.528973  ...   
4                           0.0                   0.246587  ...   
...                         ...                        ...  ...   
35035                       0.0                   0.935361  ...   
35036                       0.0                   0.538857  ...   
35037                       0.0                   0.403770  ...   
35038                       0.0                   0.436125  ...   
35039                       0.0                   0.583030  ...   

       o15 [0:vacant 1:occupied]  o16 [0:vacant 1:occupied]  \
0                       0.653947                   0.108713   
1                       

In [34]:
merged_df["timestamp [dd/mm/yyyy HH:MM]"] = pd.to_datetime(merged_df["timestamp [dd/mm/yyyy HH:MM]"], format='%d/%m/%Y %H:%M')
merged_df.set_index("timestamp [dd/mm/yyyy HH:MM]", inplace=True)
sample_df = merged_df.filter(like=" [0:vacant 1:occupied]")



In [56]:
import pandas as pd
import numpy as np
import scipy.stats as stats

# Define a function for the daily pattern
def daily_pattern(hour):
    # Mean is set to 12 for midday and standard deviation is 3
    return stats.norm.pdf(hour, 12, 3)

# Assuming your DataFrame is named sample_df and has a datetime index
for i in range(10, 25):
    sample_df[f'o{i} [0:vacant 1:occupied]'] = sample_df.index.hour.map(daily_pattern)
    
for i in range(10, 25):
    sample_df[f'o{i} [0:vacant 1:occupied]'] += np.random.normal(scale=0.1, size=len(sample_df)).round(2)
    sample_df[f'o{i} [0:vacant 1:occupied]'] = sample_df[f'o{i} [0:vacant 1:occupied]'].clip(lower=0)


C:\Users\Amghar\AppData\Local\Temp\ipykernel_56204\32447701.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sample_df[f'o{i} [0:vacant 1:occupied]'] = sample_df.index.hour.map(daily_pattern)
C:\Users\Amghar\AppData\Local\Temp\ipykernel_56204\32447701.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sample_df[f'o{i} [0:vacant 1:occupied]'] = sample_df.index.hour.map(daily_pattern)
C:\Users\Amghar\AppData\Local\Temp\ipykernel_56204\32447701.py:12: SettingWithCopyWarning: 
A value is trying to be set 

In [57]:
sample_df

ki [0:vacant 1:occupied]  \
timestamp [dd/mm/yyyy HH:MM]                             
2013-01-01 00:00:00                                0.0   
2013-01-01 00:15:00                                0.0   
2013-01-01 00:30:00                                0.0   
2013-01-01 00:45:00                                0.0   
2013-01-01 01:00:00                                0.0   
...                                                ...   
2013-12-31 22:45:00                                0.0   
2013-12-31 23:00:00                                0.0   
2013-12-31 23:15:00                                0.0   
2013-12-31 23:30:00                                0.0   
2013-12-31 23:45:00                                0.0   

                              o1_1 [0:vacant 1:occupied]  \
timestamp [dd/mm/yyyy HH:MM]                               
2013-01-01 00:00:00                                  0.0   
2013-01-01 00:15:00                                  0.0   
2013-01-01 00:30:00                                  0.0   
2013-01-01 00:45:00                                  0.0   
2013-01-01 01:00:00                                  0.0   
...                                                  ...   
2013-12-31 22:45:00                                  0.0   
2013-12-31 23:00:00                                  0.0   
2013-12-31 23:15:00                                  0.0   
2013-12-31 23:30:00                                  0.0   
2013-12-31 23:45:00                                  0.0   

                              o1_2 [0:vacant 1:occupied]  \
timestamp [dd/mm/yyyy HH:MM]                               
2013-01-01 00:00:00                                  0.0   
2013-01-01 00:15:00                                  0.0   
2013-01-01 00:30:00                                  0.0   
2013-01-01 00:45:00                                  0.0   
2013-01-01 01:00:00                                  0.0   
...                                                  ...   
2013-12-31 22:45:00                                  0.0   
2013-12-31 23:00:00                                  0.0   
2013-12-31 23:15:00                                  0.0   
2013-12-31 23:30:00                                  0.0   
2013-12-31 23:45:00                                  0.0   

                              o1_3 [0:vacant 1:occupied]  \
timestamp [dd/mm/yyyy HH:MM]                               
2013-01-01 00:00:00                                  0.0   
2013-01-01 00:15:00                                  0.0   
2013-01-01 00:30:00                                  0.0   
2013-01-01 00:45:00                                  0.0   
2013-01-01 01:00:00                                  0.0   
...                                                  ...   
2013-12-31 22:45:00                                  0.0   
2013-12-31 23:00:00                                  0.0   
2013-12-31 23:15:00                                  0.0   
2013-12-31 23:30:00                                  0.0   
2013-12-31 23:45:00                                  0.0   

                              o1_4 [0:vacant 1:occupied]  \
timestamp [dd/mm/yyyy HH:MM]                               
2013-01-01 00:00:00                                  0.0   
2013-01-01 00:15:00                                  0.0   
2013-01-01 00:30:00                                  0.0   
2013-01-01 00:45:00                                  0.0   
2013-01-01 01:00:00                                  0.0   
...                                                  ...   
2013-12-31 22:45:00                                  0.0   
2013-12-31 23:00:00                                  0.0   
2013-12-31 23:15:00                                  0.0   
2013-12-31 23:30:00                                  0.0   
2013-12-31 23:45:00                                  0.0   

                              o1_5 [0:vacant 1:occupied]  \
timestamp [dd/mm/yyyy HH:MM]                               
2013-01-01 00:00:00            

In [53]:
sample_df.to_csv("C:/Users/Amghar/Desktop/Occupancy_RandomlySimulated.csv")

In [ ]:
# Sélectionner les caractéristiques d'entrée et la variable cible
# features = ['Temperature', 'Humidity', 'AirQuality', 'Light']
# Select features and target variable

global_features = ['tempOut [C]', 'rh [%]' , 'gh [W/m2]', 'wind speed [m/s]']

#model 1 : kitchen 
features_ki = ['ki [%]', 'ki [C]', 'ki  [1:closed 0:open]', 'ki [0:off 1:on]']+ global_features
target_ki = "ki [0:vacant 1:occupied]"

# model 2 : Open desk 4  
features_o4 = ['o4 [%]', 'o4 [C]', 'o4_1 [1:closed 0:open]','o4_2 [1:closed 0:open]','o4_1 [0:off 1:on]', 'o4_2 [0:off 1:on]']+ global_features
target_o4 = "o4 [0:vacant 1:occupied]"

# model 3 : meeting room :  
features_mr = ["mr_1 [1:closed 0:open]","mr_2 [1:closed 0:open]","mr_3 [1:closed 0:open]",	"mr_4 [1:closed 0:open]", "mr_5 [1:closed 0:open]",	"mr_6 [1:closed 0:open]", "mr [C]",	"mr [%]"]+ global_features
target_mr = "mr [0:vacant 1:occupied]"

#model_4 : closed desk 3
features_o3 = ["o3_1 [1:closed 0:open]","o3_2 [1:closed 0:open]","o3_3 [1:closed 0:open]",	"o3_4 [1:closed 0:open]", "o3_1 [0:off 1:on]","o3_2 [0:off 1:on]","o3 [W]","o3 [C]","o3 [%]"]+ global_features
target_o3 = "o3 [0:vacant 1:occupied]"


| Sensor Type  | Measured Variable | Range | Accuracy | Location/Height | Functionality | Data Processed |
| ------------ | ----------------- | ----- | -------- | --------------- | ------------- | -------------- |
| Thermokon - SR04 CO2 rH | Indoor air temperature | 0–51 °C | ±1% of measuring range (typ. at 21 °C) | Inside office area, close to workstation, 0.9m height | Measures indoor air temperature | Processed by MOST building monitoring system |
| Thermokon - SR04 CO2 rH | Indoor air relative humidity | 0–100%rH | ±3% between 20–80% rH (typ. at 21 °C) | Inside office area, close to workstation, 0.9m height | Measures indoor relative humidity | Processed by MOST building monitoring system |
| Thermokon - SR04 CO2 rH | CO2 | 0–2550 ppm | ±75 ppm or ±10% of measuring range (typ.at 21 °C) | Inside office area, close to workstation, 0.9m height | Measures CO2 levels | Processed by MOST building monitoring system |
| Thermokon - SR-MDS Solar | Motion/occupancy | 0/1 | — | Ceiling-mounted | Detects movements | Processed by MOST building monitoring system |
| Thermokon - SR-MDS Solar | Brightness | 0–512Lux | — | Ceiling-mounted | Measures brightness at ceiling level | Processed by MOST building monitoring system |
| Thermokon - SRW01 | Window contact sensor | 0/1 | — | On windows | Measures the state (open/closed) of windows | Processed by MOST building monitoring system |
| Eltako - FWZ61 | Single phase energy meter | 0–3680 W | ±1% | Not specified | Measures light state indirectly | Processed by MOST building monitoring system |
| Thies Clima - Pyranometer CM3 | Solar radiation | 0–1300 W.m−2 | ±5% (350–1500 nm) | Local weather station, on top of the building, 40m above street level | Measures global radiation | Not specified |
| Thies Clima – Hygro-Thermogeber-compact 1.1005.54.000 | Outdoor air temperature | −30–+70 °C | ±0.2 K at 20 °C and wind speed >1.0 m.s−1 | Local weather station, on top of the building, 40m above street level | Measures outdoor air temperature | Not specified |
| Thies Clima – Hygro-Thermogeber-compact 1.1005.54.000 | Outdoor air relative humidity | 0–100% rH | ±2% rH | Local weather station, on top of the building, 40m above street level | Measures outdoor relative humidity | Not specified |
| Thies Clima - Windgeber-compact 4.3519.00.000 | Wind speed | 0.5–50 m.s−1 | ±0.5 m/s or ±3% of measurement | Local weather station, on top of the building, 40m above street level | Measures wind speed | Not specified |
| Thies Clima - Windrichtungsgeber- compact 4.3129.00.000 | Wind direction | 0–360° | ±5° | Local weather station, on top of the building, 40m above street level | Measures wind direction | Not specified |

In [ ]:
### Data Exploration : 

### Classification of Open Desk 4: 

Dans le cadre de notre étude sur la prévision de l'occupation des pièces, nous avons mis en œuvre plusieurs modèles pour différentes zones : la cuisine, le bureau ouvert 4, la salle de réunion et le bureau fermé 3. Chaque modèle a été élaboré à partir d'un ensemble spécifique de caractéristiques et d'une variable cible unique représentant l'occupation de la zone concernée.

Dans les données on a des varibles globales éxtrieure du batiment tel que : la température éxterieure [C], l'Humidité Exterieure [%], la radation globale [gh [W/m2]], la direction et vitesse du vent [m/s]. 

Pareil,à l'intérieure du batiement dans chaque chambre et open space on les detecteures correspandant de témperature, humidité et radation gloable. en plus, des variables binaires de tracking des ouvertures des fênetres [1/0], d'utilisation des plug électriques [1/0]. 

on défini plusieurs modéles de classification de l'occupation/absence : open Desk, Bureau fermé, salle de réuinon et cuisine. l'idée est de prédire a travers plusieurs detectors, en igorant la dimension temporelle. 

on a 4 types de chambres: Meeting Room, un bureau fermé, un bureau ouvert et une cusine. on défini 4 modéle pour chaque type, selon les variables correspandant et les variables globales; afin de voir les important caratéristiques et si sont différentes selon le type de chambre : 

Pour le bureau ouvert 4 (o4), nous avons considéré les caractéristiques suivantes :

- 'o4 [%]' : Le niveau d'humidité au bureau ouvert 4. Par exemple, un taux d'humidité élevé pourrait indiquer une présence humaine.
- 'o4 [C]' : La température au bureau ouvert 4 (en Celsius). Une température plus élevée pourrait suggérer que le bureau est occupé.
- 'o4_1 [1:fermé 0:ouvert]' et 'o4_2 [1:fermé 0:ouvert]' : L'état de deux portes ou fenêtres différentes au bureau ouvert 4, '1' indiquant fermé et '0' indiquant ouvert. Si les portes ou fenêtres sont fermées, il est possible que le bureau soit occupé.
- 'o4_1 [0:éteint 1:allumé]' et 'o4_2 [0:éteint 1:allumé]' : L'état de deux lumières ou appareils différents au bureau ouvert 4, '0' indiquant éteint et '1' indiquant allumé. Si les lumières ou appareils sont allumés, cela pourrait indiquer une présence.
De plus, chaque modèle prend en compte un ensemble de caractéristiques globales, qui sont :

- 'tempOut [C]' : La température extérieure (en Celsius). Par exemple, une température extérieure plus basse pourrait encourager plus de personnes à rester à l'intérieur.
- 'rh [%]' : Le taux d'humidité relative (%). Un taux d'humidité plus élevé à l'extérieur pourrait également inciter plus de personnes à rester à l'intérieur.
- 'gh [W/m2]' : L'énergie solaire incidente (en Watts par mètre carré). Une faible intensité lumineuse pourrait indiquer une période de la journée où l'occupation est moins probable, comme le soir ou tôt le matin.
- 'wind speed [m/s]' : La vitesse du vent (en mètres par seconde). Des conditions venteuses pourraient rendre l'extérieur moins attrayant, augmentant ainsi la probabilité d'occupation à l'intérieur.
Ces caractéristiques ont été choisies en raison de leur pertinence potentielle dans la prédiction de l'occupation. Par exemple, si les lumières sont allumées ou si les portes sont fermées, il est plus probable que le bureau soit occupé.


#### Preprocessing data : 
Sur la base de nos données, sur tous les sensors on a pas de données manquants. tous les données sont numériques; on a pas besoin de attribuer des labels ou recodées. par contre, on a un Dans l'apprentissage automatique, un problème de déséquilibre se produit lorsque les classes dans la variable cible ne sont pas représentées de manière égale. 

Par exemple, dans un problème de classification binaire, si 90% de vos données appartiennent à la Classe A et seulement 10% à la Classe B, le modèle pourrait être biaisé en faveur de la Classe A, car c'est la classe majoritaire. 
Dans notre dataset on a pas de variables catégoriques pour procéder à des techniques d'encoding aident à convertir les données catégorielles en un format numérique qui peut être utilisé par des modèles de machine learning


##### Discritization du variable Numéric d'occupation en Binaire (Présence / Absence): 
pour le cas des sensors d'occupation: 

Données Enregistrées: Une séquence de détection de mouvements d'occupants horodatés avec des valeurs binaires (indiquant la présence ou l'absence de mouvement)

Traitement des Données: Les flux de données basés sur des événements ont été traités pour générer des données d'intervalle de 15 minutes en utilisant des procédures stockées implémentées dans la base de données MySQL du système de surveillance du bâtiment MOST.

Détermination de l'Occupation: La procédure stockée dérive la durée des états d'occupation (occupé/vacant) à partir des événements stockés et renvoie l'état d'occupation dominant de chaque intervalle.

pour binner les données sert à catégoriser les valeurs de présence/absence, on prend une valeur en entrée chaque 15 min (qui représente la présence ou l'absence de mouvement). Si la valeur est 0.0, elle retourne 0, indiquant l'absence de mouvement. Pour toute autre valeur, elle retourne 1, indiquant la présence de mouvement.

##### Normalisation des variables explicatives :
La normalisation est une technique souvent appliquée dans le cadre de la préparation des données pour l'apprentissage automatique. L'objectif de la normalisation est de re-échelonner les valeurs des colonnes numériques dans l'ensemble de données sans déformer les différences dans les gammes de valeurs.

La normalisation est souvent utilisée pour garantir que toutes les données d'un ensemble de données ont la même échelle, empêchant une seule variable de dominer le modèle. Par exemple, si une variable se situe dans la plage de 0 à 1 et une autre dans la plage de 1 à 1000, le modèle pourrait être excessivement influencé par la plage plus grande. La normalisation résout ce problème en transformant toutes les variables à une échelle commune.

Il existe plusieurs méthodes courantes de normalisation, notamment :

* Normalisation Min-Max : Cette méthode rééchelonne les données à une plage fixe, généralement de 0 à 1.
* Échelonnage décimal : Cette méthode divise par la valeur absolue maximale de chaque valeur puis multiplie par 10.
* Normalisation du score Z (standardisation) : Cette méthode standardise les valeurs pour avoir une moyenne de 0 et un écart type de 1.

- pour les variables descripteurs, comme l'humidité, la temperature, la radiation globale. on utilise la standarisation pour raméne à moyenne 0 et une écart-type 1. La normalisation est une partie importante de la prétraitement des données en apprentissage automatique et peut avoir un impact significatif sur la performance d'un modèle. Cependant, elle n'est pas toujours nécessaire et peut parfois même être indésirable, selon le contexte.

##### Colinearty des variablies :

La multicolinéarité est un phénomène dans lequel une variable caractéristique dans l'ensemble de données est fortement corrélée linéairement avec une autre variable caractéristique dans le même ensemble de données. Cela peut être problématique dans un modèle de régression car cela peut exagérer l'importance des variables et rendre difficile la détermination de l'effet de chaque variable indépendamment. Il existe plusieurs façons de traiter la multicolinéarité.

Une matrice de corrélation est une table qui montre le coefficient de corrélation entre différentes variables ( 1 : une corrélation positive forte, 1 : une corrélation négative forte, 0 : l'absence de corrélation). Chaque cellule de la table montre la corrélation entre deux variables spécifiques. Une matrice de corrélation est utilisée pour résumer les données, comme une entrée dans des modèles plus avancés, et comme un outil de diagnostic pour les corrélations linéaires.

#### Compare models : 
##### Optimisation de la Précision pour Minimiser les Faux Positifs dans la Prédiction de l'Occupation des Pièces: 

Lorsque nous évoquons la notion de pénalisation des faux positifs, nous mettons l'accent sur la nécessité de minimiser les instances où notre modèle prédit de manière incorrecte la présence d'occupants. Dans notre contexte, la classe positive correspond à la "présence d'occupation", signifiant que le modèle prédit l'existence d'une présence humaine dans la pièce.

Un faux positif se produit donc lorsque le modèle anticipe une présence humaine dans la pièce (présence d'occupation), alors qu'en réalité, l'espace est vacant. Ce scénario est indésirable car il peut entraîner une consommation d'énergie superflue (par exemple, si l'éclairage ou le chauffage restent activés du fait que le système interprète la pièce comme étant occupée), ou il peut potentiellement créer des problèmes de sécurité.

Par conséquent, l'objectif est de configurer notre modèle de manière à minimiser ces prédictions de faux positifs. La précision (Precision) est un indicateur pertinent à cet égard. Elle représente le rapport des observations positives correctement prédites au total des prédictions positives. Une précision élevée indique un faible taux de faux positifs.

L'optimisation de la précision signifie que le modèle sera adapté pour améliorer sa capacité à prédire correctement la classe positive et réduire le nombre de faux positifs. Cependant, cette approche ne prend pas en compte les faux négatifs, qui, dans notre cas, se produiraient lorsque le modèle prédit que la pièce est vide, mais qu'en réalité une présence humaine est bien là. Pour prendre en compte cette situation, nous pouvons utiliser un autre indicateur, le rappel (Recall), qui mesure la capacité du modèle à trouver tous les exemples positifs.

Selon les exigences spécifiques de notre système, nous pourrions envisager d'optimiser d'autres indicateurs ou une combinaison d'indicateurs, afin d'équilibrer le compromis entre les faux positifs et les faux négatifs.

##### la définition des métriques d'Evaluation : 
- Accuracy : est la proximité des résultats de mesure avec la valeur réelle.[Wiki](https://fr.wikipedia.org/wiki/Exactitude_et_pr%C3%A9cision)

- AUC : L'aire sous le courbe ROC: caractéristique de performance (d'un test) ou courbe sensibilité/spécificité, est une mesure de la performance d'un classificateur binaire.[ROC Wiki](https://en.wikipedia.org/wiki/Receiver_operating_characteristic?oldid=375634814)

- Prec. (Precision) :  La précision montre dans quelle mesure la répétabilité (ou reproductibilité) dans des conditions inchangées affine les résultats.[Wiki](https://fr.wikipedia.org/wiki/Exactitude_et_pr%C3%A9cision)

- F1 : combine les mesures de précision et rappel, elles-mêmes basées sur les taux de vrais positifs, faux positifs et faux négatifs. L'idée de la F-mesure est de s'assurer qu'un classificateur fait de bonnes prédictions de la classe pertinente (bonne précision) en suffisamment grand nombre (bon rappel) sur un jeu de données cible. [F1 wiki](https://fr.wikipedia.org/wiki/F-mesure)

- Kappa : est la proportion de l'accord entre codeurs et Pr(e) la probabilité d'un accord aléatoire. Si les codeurs sont totalement en accord, κ = 1. S'ils sont totalement en désaccord (ou en accord dû uniquement au hasard)[Kappa Wiki](https://fr.wikipedia.org/wiki/Kappa_de_Cohen)

- MCC :Le coefficient de corrélation de Matthews (MCC, pour Matthews correlation coefficient en anglais) est une mesure de la qualité des classifications binaires. Il prend des valeurs entre -1 et +1. Un coefficient de +1 représente une prédiction parfaite, 0 une prédiction moyenne aléatoire, et -1 une prédiction inverse.Il est généralement considéré comme une mesure équilibrée car il tient compte des vrais et faux positifs et négatifs, et n'est pas trompé par les classes déséquilibrées.[Phi Coefficient](https://en.wikipedia.org/wiki/Phi_coefficient)





##### Modélisation : Classification de l'occupation :

Nous avons mis en place un modèle d'apprentissage automatique. L'objectif de ce modèle est de prédire si un bureau (désigné par 'o4') est vacant ou occupé,
en se basant sur diverses caractéristiques reliée à Open desk 4.
- Données d'origine : Le jeu de données initial comprend 35040 observations avec 11 variables.
- Données transformées : Après la transformation, le jeu de données reste de la même taille, soit 35040 observations avec 11 variables.
- Ensemble d'entraînement transformé (70%): L'ensemble d'entraînement, utilisé pour former le modèle, contient 24528 observations avec 11 variables.
- Ensemble de test transformé (30%) : L'ensemble de test, utilisé pour évaluer la performance du modèle, contient 10512 observations avec 11 variables.

Configuration du Modèle
Cible : La variable cible du modèle est 'o4', qui est une variable binaire indiquant si un emplacement est vacant (0) ou occupé (1).
Caractéristiques numériques : Le modèle utilise 10 caractéristiques numériques pour la prédiction.
Correction du déséquilibre : La technique SMOTE (Synthetic Minority Over-sampling Technique) est utilisée pour corriger le déséquilibre de classe dans le jeu de données.
Prétraitement : Aucun prétraitement automatique (comme l'encodage des variables catégorielles, l'imputation des valeurs manquantes, etc.) n'est effectué sur les données.
Normalisation : Les caractéristiques numériques sont normalisées en utilisant la méthode z-score.

##### Benchmarking des models : 

On contstate que les methodes d'ensembles d'apprentissage supérvisée est les plus performantes : les arbres de classification tel que les forêts d'arbres aléatoires, Le boosting d'arbre de classification; améliorent la qualité ou la fiabilité de la prédiction en construisant plusieurs arbres de décision depuis les données.

Définition des arbres de décision : 

C'est une technique d'apprentissage supervisé : on utilise un ensemble de données pour lesquelles on connaît la valeur de la variable-cible afin de construire l'arbre (données dites étiquetées), puis on extrapole les résultats à l'ensemble des données de test. Les arbres de décision font partie des algorithmes les plus populaires en apprentissage automatique.[Arbre de décision wiki](https://fr.wikipedia.org/wiki/Arbre_de_d%C3%A9cision_(apprentissage)#cite_note-11)

Selon les résultats de la comparaison des modèles, le "Random Forest Classifier" (Classificateur de Forêt Aléatoire) semble être le meilleur modèle avec une efficance (Accuracy) de 93% mais en termes de rappel (Recall) avec une valeur de 0.8077.

Cependant, pour déployer un modèle, il faut prendre en compte d'autres facteurs en plus de sa précision. Par exemple, le temps de formation et d'exécution, la complexité du modèle et les ressources nécessaires pour le maintenir peuvent tous affecter votre choix.

En ce qui concerne le coût en temps et en ressources, le "Decision Tree Classifier" (Classificateur d'Arbre de Décision) semble être le plus efficace, avec un temps d'exécution de seulement 0.0530 secondes. Les modèles d'arbre de décision sont également relativement simples à comprendre et à interpréter, ce qui peut faciliter leur déploiement et leur maintenance.

En revanche, le "Random Forest Classifier", bien qu'il offre une précision élevée, peut être plus coûteux en termes de temps et de ressources. Il a un temps d'exécution de 0.5520 secondes, ce qui est plus long que l'arbre de décision, et les forêts aléatoires sont également plus complexes et peuvent nécessiter plus de ressources pour le maintenir.

Il est important de noter que le meilleur choix de modèle dépendra des besoins spécifiques et des contraintes de Notre projet. Par exemple, si la précision est Notre priorité absolue, le coût supplémentaire en temps et en ressources du "Random Forest Classifier" pourrait être justifié. Si des contraintes de temps ou de ressources plus strictes, le "Decision Tree Classifier" pourrait être un meilleur choix.

##### Evaluation du modele : 
##### Interpretation et Prodcution :  
##### Resultats comparatives des autres modéles : Meeting Room,  Cuisine: 

# Evalution de test :

| model                  | Model : Top 3                           |   Accuracy |    AUC |   Recall |   Prec. |
|:-----------------------|:--------------------------------|-----------:|-------:|---------:|--------:|
| model 1 : kitchen      | Random Forest Classifier        |     0.8876 | 0.9417 |   0.7547 |  0.776  |
|       | Extra Trees Classifier          |     0.8854 | 0.9086 |   0.7499 |  0.7715 |
|      | Light Gradient Boosting Machine |     0.8812 | 0.9361 |   0.731  |  0.768  |
| model 2 : Open desk 4  | Random Forest Classifier        |     0.9338 | 0.9638 |   0.8163 |  0.8632 |
|   | Extra Trees Classifier          |     0.9323 | 0.9566 |   0.8244 |  0.8506 |
|  | Decision Tree Classifier        |     0.9127 | 0.8666 |   0.7864 |  0.798  |
| model 3 : meeting room | Light Gradient Boosting Machine |     0.9565 | 0.9802 |   0.9215 |  0.8824 |
| | Gradient Boosting Classifier    |     0.9546 | 0.9765 |   0.9137 |  0.8807 |
|  | Ada Boost Classifier            |     0.953  | 0.9742 |   0.9041 |  0.8815 |



Gh W/m2 : 0.25
o4 C : 0.16
o4 % : 0.14
o4_1 1closed 0open : 0.13
tempout C :0.11
wind speed m/s: 0.10
rh % : 0.9
o4_2 1 closed 0open: 0.01
o4_2 0off 1on : 0.009
o4_1 0off 1on : 0

|  Models                | Caractéristiques : Variables descripteurs                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | cible : Variable à prédire                                            |
|:-----------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------|
| Cuisine          | 'Humidité cuisine [%]', 'Température cuisine [C]', 'Statut porte cuisine [1:fermée 0:ouverte]', 'Statut lumière cuisine [0:éteint 1:allumé]', 'Température extérieure [C]', 'Humidité [%]', 'Radiation solaire [W/m2]', 'Vitesse du vent [m/s]'                                                                                                                                                                                                                                                                                                         | Occupation cuisine [0:vacante 1:occupée]          |
| Bureau ouvert 4  | 'Humidité bureau ouvert 4 [%]', 'Température bureau ouvert 4 [C]', 'Statut porte 1 bureau ouvert 4 [1:fermée 0:ouverte]', 'Statut porte 2 bureau ouvert 4 [1:fermée 0:ouverte]', 'Statut lumière 1 bureau ouvert 4 [0:éteint 1:allumé]', 'Statut lumière 2 bureau ouvert 4 [0:éteint 1:allumé]', 'Température extérieure [C]', 'Humidité [%]', 'Radiation solaire [W/m2]', 'Vitesse du vent [m/s]'                                                                                                                                                      | Occupation bureau ouvert 4 [0:vacant 1:occupé]    |
| Salle de réunion | 'Statut porte 1 salle de réunion [1:fermée 0:ouverte]', 'Statut porte 2 salle de réunion [1:fermée 0:ouverte]', 'Statut porte 3 salle de réunion [1:fermée 0:ouverte]', 'Statut porte 4 salle de réunion [1:fermée 0:ouverte]', 'Statut porte 5 salle de réunion [1:fermée 0:ouverte]', 'Statut porte 6 salle de réunion [1:fermée 0:ouverte]', 'Température salle de réunion [C]', 'Humidité salle de réunion [%]', 'Température extérieure [C]', 'Humidité [%]', 'Radiation solaire [W/m2]', 'Vitesse du vent [m/s]'                                  | Occupation salle de réunion [0:vacante 1:occupée] |
| Bureau fermé 3   | 'Statut porte 1 bureau fermé 3 [1:fermée 0:ouverte]', 'Statut porte 2 bureau fermé 3 [1:fermée 0:ouverte]', 'Statut porte 3 bureau fermé 3 [1:fermée 0:ouverte]', 'Statut porte 4 bureau fermé 3 [1:fermée 0:ouverte]', 'Statut lumière 1 bureau fermé 3 [0:éteint 1:allumé]', 'Statut lumière 2 bureau fermé 3 [0:éteint 1:allumé]', 'Consommation énergétique bureau fermé 3 [W]', 'Température bureau fermé 3 [C]', 'Humidité bureau fermé 3 [%]', 'Température extérieure [C]', 'Humidité [%]', 'Radiation solaire [W/m2]', 'Vitesse du vent [m/s]' | Occupation bureau fermé 3 [0:vacant 1:occupé]     |


<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>caractéristiques</th>
      <th>cible</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Cuisine</th>
      <td>[Humidité cuisine [%], Température cuisine [C], Statut porte cuisine [1:fermée 0:ouverte], Statut lumière cuisine [0:éteint 1:allumé], Température extérieure [C], Humidité [%], Radiation solaire [W/m2], Vitesse du vent [m/s]]</td>
      <td>Occupation cuisine [0:vacante 1:occupée]</td>
    </tr>
    <tr>
      <th>Bureau ouvert 4</th>
      <td>[Humidité bureau ouvert 4 [%], Température bureau ouvert 4 [C], Statut porte 1 bureau ouvert 4 [1:fermée 0:ouverte], Statut porte 2 bureau ouvert 4 [1:fermée 0:ouverte], Statut lumière 1 bureau ouvert 4 [0:éteint 1:allumé], Statut lumière 2 bureau ouvert 4 [0:éteint 1:allumé], Température extérieure [C], Humidité [%], Radiation solaire [W/m2], Vitesse du vent [m/s]]</td>
      <td>Occupation bureau ouvert 4 [0:vacant 1:occupé]</td>
    </tr>
    <tr>
      <th>Salle de réunion</th>
      <td>[Statut porte 1 salle de réunion [1:fermée 0:ouverte], Statut porte 2 salle de réunion [1:fermée 0:ouverte], Statut porte 3 salle de réunion [1:fermée 0:ouverte], Statut porte 4 salle de réunion [1:fermée 0:ouverte], Statut porte 5 salle de réunion [1:fermée 0:ouverte], Statut porte 6 salle de réunion [1:fermée 0:ouverte], Température salle de réunion [C], Humidité salle de réunion [%], Température extérieure [C], Humidité [%], Radiation solaire [W/m2], Vitesse du vent [m/s]]</td>
      <td>Occupation salle de réunion [0:vacante 1:occupée]</td>
    </tr>
    <tr>
      <th>Bureau fermé 3</th>
      <td>[Statut porte 1 bureau fermé 3 [1:fermée 0:ouverte], Statut porte 2 bureau fermé 3 [1:fermée 0:ouverte], Statut porte 3 bureau fermé 3 [1:fermée 0:ouverte], Statut porte 4 bureau fermé 3 [1:fermée 0:ouverte], Statut lumière 1 bureau fermé 3 [0:éteint 1:allumé], Statut lumière 2 bureau fermé 3 [0:éteint 1:allumé], Consommation énergétique bureau fermé 3 [W], Température bureau fermé 3 [C], Humidité bureau fermé 3 [%], Température extérieure [C], Humidité [%], Radiation solaire [W/m2], Vitesse du vent [m/s]]</td>
      <td>Occupation bureau fermé 3 [0:vacant 1:occupé]</td>
    </tr>
  </tbody>
</table>
